File: data/src/main/java/org/akvo/flow/data/repository/FormDataRepository.java
/*
 * Copyright (C) 2018-2020 Stichting Akvo (Akvo Foundation)
 *
 * This file is part of Akvo Flow.
 *
 * Akvo Flow is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Akvo Flow is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Akvo Flow.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.akvo.flow.data.repository;

import android.database.Cursor;

import org.akvo.flow.data.datasource.DataSourceFactory;
import org.akvo.flow.data.datasource.DatabaseDataSource;
import org.akvo.flow.data.entity.ApiFormHeader;
import org.akvo.flow.data.entity.form.DomainFormMapper;
import org.akvo.flow.data.entity.form.Form;
import org.akvo.flow.data.entity.form.FormHeaderParser;
import org.akvo.flow.data.entity.form.FormIdMapper;
import org.akvo.flow.data.entity.form.XmlFormParser;
import org.akvo.flow.data.net.RestApi;
import org.akvo.flow.data.net.s3.S3RestApi;
import org.akvo.flow.data.util.FlowFileBrowser;
import org.akvo.flow.domain.entity.DomainForm;
import org.akvo.flow.domain.repository.FormRepository;
import org.jetbrains.annotations.NotNull;

import java.io.InputStream;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import okhttp3.ResponseBody;
import timber.log.Timber;

public class FormDataRepository implements FormRepository {

    private static final String TEST_FORM_ID = "0";

    private final FormHeaderParser formHeaderParser;
    private final XmlFormParser xmlParser;
    private final RestApi restApi;
    private final DataSourceFactory dataSourceFactory;
    private final FormIdMapper formIdMapper;
    private final S3RestApi s3RestApi;
    private final DomainFormMapper domainFormMapper;

    @Inject
    public FormDataRepository(FormHeaderParser formHeaderParser, XmlFormParser xmlParser,
                              RestApi restApi, DataSourceFactory dataSourceFactory, FormIdMapper formIdMapper,
                              S3RestApi s3RestApi, DomainFormMapper domainFormMapper) {
        this.formHeaderParser = formHeaderParser;
        this.xmlParser = xmlParser;
        this.restApi = restApi;
        this.dataSourceFactory = dataSourceFactory;
        this.formIdMapper = formIdMapper;
        this.s3RestApi = s3RestApi;
        this.domainFormMapper = domainFormMapper;
    }

    @Override
    public Observable<Boolean> loadForm(String formId, String deviceId) {
        final DatabaseDataSource dataBaseDataSource = dataSourceFactory.getDataBaseDataSource();
        if (TEST_FORM_ID.equals(formId)) {
            return dataBaseDataSource.installTestForm();
        } else {
            return downloadFormHeader(formId, deviceId);
        }
    }

    @Override
    public Observable<Integer> reloadForms(final String deviceId) {
        final DatabaseDataSource dataBaseDataSource = dataSourceFactory.getDataBaseDataSource();
        return dataBaseDataSource.getFormIds()
                .map(new Function<Cursor, List<String>>() {
                    @Override
                    public List<String> apply(Cursor cursor) {
                        return formIdMapper.mapToFormId(cursor);
                    }
                })
                .concatMap(new Function<List<String>, Observable<Integer>>() {
                    @Override
                    public Observable<Integer> apply(final List<String> formIds) {
                        return dataBaseDataSource.deleteAllForms()
                                .concatMap(new Function<Boolean, Observable<Integer>>() {
                                    @Override
                                    public Observable<Integer> apply(Boolean aBoolean) {
                                        return downloadFormHeaders(formIds, deviceId);
                                    }
                                });
                    }
                });
    }

    @Override
    public Observable<Integer> downloadForms(final String deviceId) {
        return restApi.downloadFormsHeader(deviceId)
                .map(new Function<String, List<ApiFormHeader>>() {
                    @Override
                    public List<ApiFormHeader> apply(String response) {
                        return formHeaderParser.parseMultiple(response);
                    }
                })
                .concatMap(new Function<List<ApiFormHeader>, Observable<Integer>>() {
                    @Override
                    public Observable<Integer> apply(List<ApiFormHeader> apiFormHeaders) {
                        return downloadForms(apiFormHeaders);
                    }
                });
    }

    @Override
    @NotNull
    public DomainForm getForm(@NotNull String formId) {
        return domainFormMapper.mapForm(dataSourceFactory.getDataBaseDataSource().getForm(formId));
    }

    @NotNull
    @Override
    public List<DomainForm> getForms(long surveyId) {
        return domainFormMapper.mapForms(dataSourceFactory.getDataBaseDataSource().getForms(surveyId));
    }

    private Observable<Boolean> downloadFormHeader(String formId, String deviceId) {
        return restApi.downloadFormHeader(formId, deviceId)
                .map(new Function<String, ApiFormHeader>() {
                    @Override
                    public ApiFormHeader apply(String response) {
                        return formHeaderParser.parseOne(response);
                    }
                })
                .concatMap(new Function<ApiFormHeader, Observable<Boolean>>() {
                    @Override
                    public Observable<Boolean> apply(final ApiFormHeader apiFormHeader) {
                        return insertAndDownload(apiFormHeader);
                    }
                });
    }

    private Observable<Integer> downloadForms(List<ApiFormHeader> apiFormHeaders) {
        return Observable.fromIterable(apiFormHeaders)
                .concatMap(new Function<ApiFormHeader, Observable<Boolean>>() {
                    @Override
                    public Observable<Boolean> apply(final ApiFormHeader apiFormHeader) {
                        return insertAndDownload(apiFormHeader);

                    }
                })
                .toList()
                .toObservable()
                .map(new Function<List<Boolean>, Integer>() {
                    @Override
                    public Integer apply(List<Boolean> booleans) {
                        return booleans.size();
                    }
                });
    }

    private Observable<Boolean> insertAndDownload(final ApiFormHeader apiFormHeader) {
        return dataSourceFactory.getDataBaseDataSource().insertSurveyGroup(apiFormHeader)
                .concatMap(new Function<Boolean, Observable<Boolean>>() {
                    @Override
                    public Observable<Boolean> apply(Boolean ignored) {
                        return downloadForm(apiFormHeader);
                    }
                });
    }

    private Observable<Integer> downloadFormHeaders(List<String> formIds, final String deviceId) {
        return Observable.fromIterable(formIds)
                .concatMap(new Function<String, Observable<Boolean>>() {
                    @Override
                    public Observable<Boolean> apply(String formId) {
                        return downloadFormHeader(formId, deviceId);
                    }
                })
                .toList()
                .toObservable()
                .map(new Function<List<Boolean>, Integer>() {
                    @Override
                    public Integer apply(List<Boolean> booleans) {
                        return booleans.size();
                    }
                });
    }

    private Observable<Boolean> downloadForm(final ApiFormHeader apiFormHeader) {
        return dataSourceFactory.getDataBaseDataSource().formNeedsUpdate(apiFormHeader)
                .concatMap(new Function<Boolean, Observable<Boolean>>() {
                    @Override
                    public Observable<Boolean> apply(Boolean updateNeeded) {
                        if (updateNeeded) {
                            return downloadAndSaveForm(apiFormHeader);
                        } else {
                            return Observable.just(true);
                        }
                    }
                });
    }

    private Observable<Boolean> downloadAndSaveForm(final ApiFormHeader apiFormHeader) {
        return downloadAndExtractFile(apiFormHeader.getId() + FlowFileBrowser.ZIP_SUFFIX,
                FlowFileBrowser.DIR_FORMS)
                .concatMap((Function<Boolean, Observable<Boolean>>) aBoolean -> saveForm(apiFormHeader)
                        .concatMap((Function<Boolean, Observable<Boolean>>) aBoolean1 -> {
                            dataSourceFactory.getDataBaseDataSource().setSurveyUnViewed(apiFormHeader.getGroupId());
                            return Observable.just(true);
                        }));
    }

    private Observable<Boolean> downloadAndExtractFile(final String fileName, final String folder) {
        return s3RestApi.downloadArchive(fileName)
                .concatMap(new Function<ResponseBody, Observable<Boolean>>() {
                    @Override
                    public Observable<Boolean> apply(ResponseBody responseBody) {
                        return dataSourceFactory.getFileDataSource()
                                .extractRemoteArchive(responseBody, folder);

                    }
                });
    }

    private Observable<Boolean> saveForm(final ApiFormHeader apiFormHeader) {
        return dataSourceFactory.getFileDataSource().getFormFile(apiFormHeader.getId())
                .map(new Function<InputStream, Form>() {
                    @Override
                    public Form apply(InputStream inputStream) {
                        return xmlParser.parse(inputStream);
                    }
                })
                .concatMap(new Function<Form, Observable<Boolean>>() {
                    @Override
                    public Observable<Boolean> apply(final Form form) {
                        return downloadResources(form)
                                .concatMap(new Function<Boolean, Observable<Boolean>>() {
                                    @Override
                                    public Observable<Boolean> apply(Boolean aBoolean) {
                                        return dataSourceFactory.getDataBaseDataSource()
                                                .insertSurvey(apiFormHeader, true, form);
                                    }
                                })
                                .doOnError(new Consumer<Throwable>() {
                                    @Override
                                    public void accept(Throwable throwable) {
                                        Timber.e(throwable);
                                        dataSourceFactory.getDataBaseDataSource()
                                                .insertSurvey(apiFormHeader, false, form);
                                    }
                                });
                    }
                });

    }

    private Observable<Boolean> downloadResources(Form form) {
        return Observable.fromIterable(form.getResources())
                .concatMap(new Function<String, Observable<Boolean>>() {
                    @Override
                    public Observable<Boolean> apply(String resource) {
                        return downloadAndExtractFile(resource + FlowFileBrowser.ZIP_SUFFIX,
                                FlowFileBrowser.DIR_RES);
                    }
                })
                .toList()
                .toObservable()
                .map(new Function<List<Boolean>, Boolean>() {
                    @Override
                    public Boolean apply(List<Boolean> ignored) {
                        return true;
                    }
                });
    }
}

File: data/src/main/java/org/akvo/flow/data/repository/FormDataRepository.kt
/*
 * Copyright (C) 2018-2020 Stichting Akvo (Akvo Foundation)
 *
 * This file is part of Akvo Flow.
 *
 * Akvo Flow is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Akvo Flow is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Akvo Flow.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.akvo.flow.data.repository

import io.reactivex.Observable
import org.akvo.flow.data.datasource.DataSourceFactory
import org.akvo.flow.data.entity.ApiFormHeader
import org.akvo.flow.data.entity.form.DataForm
import org.akvo.flow.data.entity.form.DataFormMapper
import org.akvo.flow.data.entity.form.DataSurveyMapper
import org.akvo.flow.data.entity.form.DomainFormMapper
import org.akvo.flow.data.entity.form.FormHeaderParser
import org.akvo.flow.data.entity.form.FormIdMapper
import org.akvo.flow.data.net.RestApi
import org.akvo.flow.data.net.s3.S3RestApi
import org.akvo.flow.data.util.FlowFileBrowser.DIR_FORMS
import org.akvo.flow.data.util.FlowFileBrowser.DIR_RES
import org.akvo.flow.data.util.FlowFileBrowser.ZIP_SUFFIX
import org.akvo.flow.domain.entity.DomainForm
import org.akvo.flow.domain.exception.CascadeProcessingError
import org.akvo.flow.domain.exception.WrongDashboardError
import org.akvo.flow.domain.interactor.bootstrap.BootstrapProcessor
import org.akvo.flow.domain.repository.FormRepository
import org.akvo.flow.utils.XmlFormParser
import org.akvo.flow.utils.entity.Question
import timber.log.Timber
import java.io.File
import java.util.Enumeration
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import javax.inject.Inject


class FormDataRepository @Inject constructor(
    private val formHeaderParser: FormHeaderParser,
    private val xmlParser: XmlFormParser,
    private val restApi: RestApi,
    private val dataSourceFactory: DataSourceFactory,
    private val formIdMapper: FormIdMapper,
    private val s3RestApi: S3RestApi,
    private val domainFormMapper: DomainFormMapper,
    private val dataFormMapper: DataFormMapper,
    private val dataSurveyMapper: DataSurveyMapper
) : FormRepository {

    override fun loadForm(formId: String?, deviceId: String?): Observable<Boolean?>? {
        val dataBaseDataSource = dataSourceFactory.dataBaseDataSource
        return if (TEST_FORM_ID == formId) {
            dataBaseDataSource.installTestForm()
        } else {
            downloadFormHeader(formId, deviceId)
        }
    }

    override fun reloadForms(deviceId: String?): Observable<Int?>? {
        val dataBaseDataSource = dataSourceFactory.dataBaseDataSource
        return dataBaseDataSource.formIds
            .map { cursor -> formIdMapper.mapToFormId(cursor) }
            .concatMap { formIds ->
                dataBaseDataSource.deleteAllForms()
                    .concatMap { downloadFormHeaders(formIds, deviceId) }
            }
    }

    override fun downloadForms(deviceId: String?): Observable<Int?>? {
        return restApi.downloadFormsHeader(deviceId)
            .map { response -> formHeaderParser.parseMultiple(response) }
            .concatMap { apiFormHeaders -> downloadForms(apiFormHeaders) }
    }

    override suspend fun loadFormLanguages(formId: String): Set<String> {
        return xmlParser.parseLanguages(dataSourceFactory.fileDataSource.getFormFile(formId))
    }

    override fun getForm(formId: String): DomainForm {
        return domainFormMapper.mapForm(dataSourceFactory.dataBaseDataSource.getForm(formId))
    }

    override fun getForms(surveyId: Long): List<DomainForm> {
        return domainFormMapper.mapForms(dataSourceFactory.dataBaseDataSource.getForms(surveyId))
    }

    override suspend fun getFormWithGroups(formId: String): DomainForm {
        //TODO: once questions are in database we need to parse from db everything without opening xml form
        return domainFormMapper.mapForm(
            dataSourceFactory.dataBaseDataSource.getFormWithGroups(formId),
            parseFormQuestions(formId)
        )
    }

    override suspend fun processZipFile(file: File, instanceUrl: String, awsBucket: String) {
        val zipFile = ZipFile(file)
        val entries: Enumeration<out ZipEntry> = zipFile.entries()
        while (entries.hasMoreElements()) {
            val entry = entries.nextElement()
            val entryName = entry.name ?: ""
            when {
                entryName.endsWith(BootstrapProcessor.CASCADE_RES_SUFFIX) -> {
                    processCascadeResource(zipFile, entry)
                }
                entryName.endsWith(BootstrapProcessor.XML_SUFFIX) -> {
                    processSurveyFile(zipFile, entry, instanceUrl, awsBucket)
                }
            }
        }
        zipFile.close()
        file.renameTo(File(file.absolutePath + BootstrapProcessor.PROCESSED_OK_SUFFIX))
    }

    private fun processSurveyFile(
        zipFile: ZipFile,
        entry: ZipEntry,
        instanceUrl: String,
        awsBucket: String,
    ) {
        val inputStream = zipFile.getInputStream(entry)
        val formAndMeta = dataFormMapper.mapFormAndMetadata(xmlParser.parseXmlFormWithMeta(inputStream))

        //verify form
        val surveyMetadata = formAndMeta.second
        if (surveyMetadata.alias.isNotEmpty()) {
            if (!instanceUrl.contains(surveyMetadata.alias)) {
                throw WrongDashboardError()
            }
        } else if (surveyMetadata.app.isNotEmpty()) {
            if (!awsBucket.contains(surveyMetadata.app)) {
                throw WrongDashboardError()
            }
        } else {
            throw WrongDashboardError()
        }

        //extract form to app folder
        dataSourceFactory.fileDataSource.copyFormFile(zipFile, entry, formAndMeta.first.formId)
        dataSourceFactory.dataBaseDataSource.insertSurveyGroup(surveyMetadata.survey)
        saveFormAndGroups(formAndMeta.first, true)
    }

    private fun processCascadeResource(zipFile: ZipFile, entry: ZipEntry) {
        return try {
            dataSourceFactory.fileDataSource.extractZipEntry(zipFile, entry, DIR_RES)
        } catch (e: Exception) {
            Timber.e(e)
            throw CascadeProcessingError(e)
        }
    }

    private fun parseFormQuestions(formId: String): HashMap<Int, MutableList<Question>> {
        return xmlParser.parseXmlQuestions(dataSourceFactory.fileDataSource.getFormFile(formId))
    }

    private fun downloadFormHeader(formId: String?, deviceId: String?): Observable<Boolean?> {
        return restApi.downloadFormHeader(formId!!, deviceId)
            .map { response -> formHeaderParser.parseOne(response) }
            .concatMap { apiFormHeader -> insertAndDownload(apiFormHeader) }
    }

    private fun downloadForms(apiFormHeaders: List<ApiFormHeader>): Observable<Int> {
        return Observable.fromIterable(apiFormHeaders)
            .concatMap { apiFormHeader -> insertAndDownload(apiFormHeader) }
            .toList()
            .toObservable()
            .map { booleans -> booleans.size }
    }

    private fun insertAndDownload(apiFormHeader: ApiFormHeader): Observable<Boolean> {
        val map = dataSurveyMapper.map(apiFormHeader)
        return dataSourceFactory.dataBaseDataSource.insertSurveyGroup(map)
            .concatMap { downloadForm(apiFormHeader) }
    }

    private fun downloadFormHeaders(formIds: List<String>, deviceId: String?): Observable<Int> {
        return Observable.fromIterable(formIds)
            .concatMap<Boolean> { formId -> downloadFormHeader(formId, deviceId) }
            .toList()
            .toObservable()
            .map { booleans -> booleans.size }
    }

    private fun downloadForm(apiFormHeader: ApiFormHeader): Observable<Boolean> {
        return dataSourceFactory.dataBaseDataSource.formNeedsUpdate(apiFormHeader)
            .concatMap { updateNeeded ->
                if (updateNeeded) {
                    downloadAndSaveForm(apiFormHeader)
                } else {
                    Observable.just(true)
                }
            }
    }

    private fun downloadAndSaveForm(apiFormHeader: ApiFormHeader): Observable<Boolean> {
        return downloadAndExtractFile(apiFormHeader.id + ZIP_SUFFIX, DIR_FORMS)
            .concatMap { saveForm(apiFormHeader) }
    }

    private fun downloadAndExtractFile(
        fileName: String,
        folder: String,
    ): Observable<Boolean> {
        return s3RestApi.downloadArchive(fileName)
            .concatMap { responseBody ->
                dataSourceFactory.fileDataSource.extractRemoteArchive(responseBody, folder)
            }
    }

    private fun saveForm(apiFormHeader: ApiFormHeader): Observable<Boolean> {
        val inputStream = dataSourceFactory.fileDataSource.getFormFile(apiFormHeader.id)
        val backUpVersion = apiFormHeader.version.toDouble()
        val form = dataFormMapper.mapForm(xmlParser.parseXmlForm(inputStream, backUpVersion))
        return downloadResources(form)
            .concatMap {
                saveFormAndGroups(form, true)
            }
            .doOnError { throwable ->
                Timber.e(throwable)
                saveFormAndGroups(form, false)
            }
    }

    private fun saveFormAndGroups(form: DataForm, resourcesDownloaded: Boolean): Observable<Boolean> {
        val dataBaseDataSource = dataSourceFactory.dataBaseDataSource
        dataBaseDataSource.saveForm(resourcesDownloaded, form)
        dataBaseDataSource.saveQuestionGroups(form)
        return Observable.just(true)
    }

    private fun downloadResources(form: DataForm): Observable<Boolean> {
        return Observable.fromIterable(form.getResources())
            .concatMap { resource ->
                downloadAndExtractFile(resource + ZIP_SUFFIX, DIR_RES)
            }
            .toList()
            .toObservable()
            .map { true }
    }

    companion object {
        private const val TEST_FORM_ID = "0"
    }

}